1-Draft/RFCXXXX-Umask-and-Ulimit.md

```diff
@@ -0,0 +1,276 @@
+GetHashCode Method     int GetHashCode()
+GetType     Method     type GetType()
+ToString    Method     string ToString()
+Current     Property   ulong Current {get;set;}
+Maximum     Property   ulong Maximum {get;set;}
+Resource    Property   System.Management.Automation.Platform+ResourceForUlimit Resource {get;set;}
+```
+
+#### Implementation Notes
+
+The `ulimit` cmdlets do not totally replicate the Unix behavior.
+Specifically, resources which are not alterable via the `setrlimit` API will not be supported.
+The pipe size resource is not retrievable or alterable with the ulimit cmdlets.
+The resources available will vary based on the platform, the following table shows what resources shall be returned based on their platform:
+
+| Resource | MacOS | Linux |
+| -------- | :---: | :---: |
+| Maximum CPU Time | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum File Size | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum Data Segment Size | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum Stack Size | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum Core File Size | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum Resident Set Size | :heavy_check_mark: | :heavy_check_mark: |
+| Memory Address Space | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum number of processes | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum number of open Files | :heavy_check_mark: | :heavy_check_mark: |
+| Maximum locked-in-memory address space | :heavy_check_mark: | :heavy_check_mark: |
+| Address space limit | | :heavy_check_mark: |
+| Maximum number of file locks | | :heavy_check_mark: |
+| Maximum number of pending signals | | :heavy_check_mark: |
+| Maximum bytes in message queue | | :heavy_check_mark: |
+| Maximum nice priority allowed | | :heavy_check_mark: |
+| Maximum real-time priority | | :heavy_check_mark: |
+| Timeout for real-time tasks (microseconds) | | :heavy_check_mark: |
+
+#### Set-Ulimit Syntax
+
+```powershell
+NAME
+    Set-Ulimit
+
+SYNTAX
+    Set-Ulimit [-ResourceInfo] <Platform+ResourceLimitInfo> [<CommonParameters>]
+
+    Set-Ulimit -Resource {CpuTime | FileSize | DataSegmentSize | StackSize | CoreFileSize | AddresSpaceSize | MemoryAddressSize | 
+    NumberOfProcesses | NumberOfOpenFiles} [-Current <ulong>] [-Maximum <ulong>] [<CommonParameters>]
+
+ALIASES
+    None
+```
+
+#### Set-Ulimit Examples
+
+```powershell
+PS> get-ulimit -Resource NumberOfOpenFiles
+
+         Resource Current             Maximum
+         -------- -------             -------
+NumberOfOpenFiles   10240 9223372036854775807
+
+PS> set-ulimit -Resource NumberOfOpenFiles -Current 20kb 
+PS> get-ulimit -Resource NumberOfOpenFiles 
+
+         Resource Current             Maximum
+         -------- -------             -------
+NumberOfOpenFiles   20480 9223372036854775807
+```
+
+# Alternative and Additional Proposals
+
+## Umask behavior
+
+A function `Umask` may be written to mimic the behavior of the Unix umask utility.
```

Comment:
Note that _not_ writing such functions (`umask`, `ulimit`) will implicitly invoke the `Get-UMask` and `Get-Ulimit` cmdlets when the user submits `umask` and `ulimit` on Linux (where no external-utility implementations exists), due to PowerShell falling back to a `Get`-prefixed command (default-verb logic) if no command is found with the exact name specified.  

Also, in the case of `umask`, Unix users will likely expect support for the single-letter options supported by the Bash builtin, such as `-n` for `-Resource NumberOfOpenFiles `

Reply:
In fact we discussed exactly this. The expectation of bash-compatible options (and the difficulty of implementing those as a PowerShell function) seems to motivate against it

Reply:
@rjmholt, I agree that it's generally not worth emulating the CLI options of other shell's builtins (or a platform's standard utilities).
The tricky thing here is that we have an unfortunate name collision on Linux, where executing `umask` will, in fact, invoke `Get-UMask` (due to the absence of an external `umask` command, combined with the default-verb logic), which will result in output that differs from Bash's `umask` - but perhaps that won't be a big problem in practice.